core/string/ustring.cpp

```diff
@@ -4545,84 +4545,128 @@ bool String::is_network_share_path() const {
 	if (p > 0) {
 		bool only_chars = true;
 		for (int i = 0; i < p; i++) {
-			if (!is_ascii_alphanumeric_char(s[i])) {
+			if (!is_ascii_alphanumeric_char(p_str[i])) {
 				only_chars = false;
 				break;
 			}
 		}
 		if (only_chars) {
-			found = true;
-			drive = s.substr(0, p + 3);
-			s = s.substr(p + 3);
-		}
-	}
-	if (!found) {
-		if (is_network_share_path()) {
-			// Network path, beginning with // or \\.
-			drive = s.substr(0, 2);
-			s = s.substr(2);
-		} else if (s.begins_with("/") || s.begins_with("\\")) {
-			// Absolute path.
-			drive = s.substr(0, 1);
-			s = s.substr(1);
-		} else {
-			// Windows-style drive path, like C:/ or C:\.
-			p = s.find(":/");
-			if (p == -1) {
-				p = s.find(":\\");
-			}
-			if (p != -1 && p < s.find_char('/')) {
-				drive = s.substr(0, p + 2);
-				s = s.substr(p + 2);
-			}
+			return p + 3;
 		}
 	}
 
-	s = s.replace("\\", "/");
-	while (true) { // in case of using 2 or more slash
-		String compare = s.replace("//", "/");
-		if (s == compare) {
-			break;
-		} else {
-			s = compare;
-		}
+	if (p_str.is_network_share_path()) {
+		// Network path, beginning with // or \\.
+		return 2;
+	}
+
+	if (p_str.begins_with("/") || p_str.begins_with("\\")) {
+		// Absolute path.
+		return 1;
 	}
-	Vector<String> dirs = s.split("/", false);
 
-	for (int i = 0; i < dirs.size(); i++) {
-		String d = dirs[i];
-		if (d == ".") {
-			dirs.remove_at(i);
-			i--;
-		} else if (d == "..") {
-			if (i != 0) {
-				dirs.remove_at(i);
-				dirs.remove_at(i - 1);
-				i -= 2;
+	// Windows-style drive path, like C:/ or C:\.
+	p = p_str.find(":/");
+	if (p == -1) {
+		p = p_str.find(":\\");
+	}
+	if (p != -1) {
+		bool no_separator = true;
+		const char32_t *ptr = p_str.ptr();
+		for (const char32_t *end = ptr + p; ptr < end; ptr++) {
+			if (is_path_seperator(*ptr)) {
+				no_separator = false;
+				break;
 			}
 		}
+		if (no_separator) {
+			return p + 2;
+		}
 	}
 
-	s = "";
+	return 0;
+}
 
-	for (int i = 0; i < dirs.size(); i++) {
-		if (i > 0) {
-			s += "/";
+String String::simplify_path() const {
+	if (is_empty()) {
+		return *this;
+	}
+
+	String ret;
+	// The new string is guaranteed to be equal to or shorter than us.
+	// If shorter, it will be shrunk at the end of the function.
+	ret.resize(size());
+
+	// dst points to the next character to write.
+	char32_t *dst = ret.ptrw();
+
+	// src points to the next character being read.
+	const char32_t *src = ptr();
+	const char32_t *end = src + length();
+
+	// Consume drive part verbatim.
+	for (int drive_end_idx = find_drive_end_idx(*this); drive_end_idx > 0; --drive_end_idx) {
+		const char32_t ch = *src++;
+		*dst++ = ch == '\\' ? '/' : ch;
+	}
+
+	// Part of dst beyond which we don't backtrack, i.e. drive and potential ../ parts.
+	char32_t *dst_fixed = dst;
+	const char32_t *dst_path_start = dst;
+
+	for (; src < end; ++src) {
+		if (is_path_seperator(*src)) {
+			// Double slash; ignore.
+			continue;
+		}
+
+		char32_t *dst_part_start = dst;
```

Comment:
so all this backtracking part seems unecessary. If I understand correctly (which I may not) we just are looking for special tokens (ie : `./` and `../`). If it is `./` we just skip it basically, without copying it into dest. If it is `../` we just want to remove the portion until the last `/` we found ie ( `C:/super/../` becomes `C:/`). The first case definetly does not need backtracking. You can just compare `if ((*src == '.') && ((src+1) == end || is_path_seperator(*src + 1)) src += 2` which involves no copy. The second one `if ((*src == '.') && ((src+1) < end && (*(src + 1) == '.')  && ((src+2) == end || is_path_seperator(*src + 2))) src += 3` , and you need to save a char* to the last `/` you appended and reset dst to this. If you saved no char yet, then you are in the `// Can't backtrack, so save double dot in the fixed section.`, and you copy them in. If not in those 2 cases, then copy until last separator as you already do. It will save  you some copies.

Reply:
This logic would not work with paths like `path/to/../../`. To avoid copying you will need to forward check the entire string for each path element, to see if it gets removed in the end. This would change its runtime to `O(n log n)` from `O(n)` so it's simply not worth it.

Reply:
I don't think the complexity will change, but you have a point that you need to store all path separators locations for it to work. If this implies a memory allocation, then there is no point in doing it. Could imagine a static array, but then when a demonic scripts give you back a 2mb path consisting only in `/../../` we will overflow. So I would agree with you to keep it as is.